### PR TITLE
Add model-monitoring stack and update model-mesh manifests to latest

### PR DESCRIPTION
**Pull request by VedantMahabaleshwarkar**
Model-Mesh updates + add model monitoring stack

## Description
- Sync model-mesh manifests from opendatahub-io/model-serving/manifests to opendatahub-io/odh-manifests/model-mesh
- Add model-monitoring manifests from downstream manifests to ODH manifests
- Delete old modelmesh basic tests that are broken
- Add following tests to `tests/basictests/modelmesh.sh`
  - Resource creation
    - Verify etcd deployment succeeded
    - Verify etcd secrets were created successfully
    - Verify odh-prometheus-operator deployment was successful
    - Verify 3 prometheus pods came up successfully
    - Verify odh-model-monitoring route was created successfully
  - Setup model testing namespace
    - Deleted old sklearn model yaml files and deployment since it was broken
    - Added new model creation files from [model-serving quickstart](https://github.com/opendatahub-io/modelmesh-serving/tree/main/quickstart)
  - Inference Testing
    - Verify curl call to model is successful
  - Model Monitoring
    - Verify accurate number of models deployed from openshift-monitoring (old test kept mostly as it was)
    - Verify new model-monitoring prometheus stack works by making a prometheus query against the `odh-model-monitoring` route and verify result

## How Has This Been Tested?
- Install a fresh cluster
- Install ODH operator
- Create namespace where you want to install the ODH components. eg : `opendatahub`
- Use the following kfdef to install the odh-common components + model-mesh + model-monitoring
- Edit the kfdef to replace `monitoring-namespace` and `deployment-namespace` with the NS you are installing ODH components in (if it is not `opendatahub`)
```
apiVersion: kfdef.apps.kubeflow.org/v1
kind: KfDef
metadata:
  name: odh-modelmesh
spec:
  applications:
    - kustomizeConfig:
        repoRef:
          name: manifests
          path: odh-common
      name: odh-common
    - kustomizeConfig:
        parameters:
        - name: monitoring-namespace
          value: opendatahub
        overlays:
          - odh-model-controller
        repoRef:
          name: manifests
          path: model-mesh
      name: model-mesh
    - kustomizeConfig:
        parameters:
        - name: deployment-namespace
          value: opendatahub
        repoRef:
          name: manifests
          path: modelmesh-monitoring
      name: modelmesh-monitoring
  repos:
    - name: manifests
      uri: https://api.github.com/repos/VedantMahabaleshwarkar/odh-manifests/tarball/modelmesh
  version: master 
 ```
- Verify all deployments are successful with all containers coming up successfully
- Run the modelmesh test suite as follows
  - modify `tests/makefile`
    - GIT_ORG=VedantMahabaleshwarkar
    - GIT_BRANCH=modelmesh
  - in `odh-manifests/tests`
    - make build
    - make run SKIP_INSTALL=true SKIP_KFDEF_INSTALL=true TESTS_REGEX=modelmesh (we already installed the operator and installed a kfdef earlier in the test instructions).    

#### Note:
Overall test suite fails due to `/root/peak/util: line 23: orig_project: unbound variable` but verify that all modelmesh tests passed  

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by LaVLaS** [comment]
/hold

Putting a hold on this until after we tag odh-manifests for the v1.4.2 release today

**Comment by LaVLaS** [comment]
@VedantMahabaleshwarkar Since this is adding new functionality, it's going to need some type a smoke test update to verify it's actually working

**Comment by VedantMahabaleshwarkar** [comment]
> @VedantMahabaleshwarkar Since this is adding new functionality, it's going to need some type a smoke test update to verify it's actually working

Added tests

**Comment by VedantMahabaleshwarkar** [comment]
> @VedantMahabaleshwarkar Since this is adding new functionality, it's going to need some type a smoke test update to verify it's actually working

@LaVLaS Added tests to the PR as well

**Comment by VannTen** [comment]
>     * Sync model-mesh manifests from opendatahub-io/model-serving/manifests to opendatahub-io/odh-manifests/model-mesh
> 
>     * Add model-monitoring manifests from downstream manifests to ODH manifests
> 
>     * Delete old modelmesh basic tests that are broken
> 
>     * Add following tests to `tests/basictests/modelmesh.sh

Is there some  circular dependency between those items ?
Looks like that could be broken into separate PR for easier review.

**Review by VannTen — DISMISSED** [review]
@LaVLaS 

I took a quick look.
Some stuff looks a bit alarming to me, see below.

File: modelmesh-monitoring/base/servicemonitors/modelmesh-federated-metrics.yaml
Comment (VedantMahabaleshwarkar) [reply modelmesh-monitoring/base/servicemonitors/modelmesh-federated-metrics.yaml:23]: This is for the benefit of the Dashboard. The plan is that the odh-dashboard will show some of these system metrics in the dashboard. We are federating these specific metrics from `openshift-monitoring` so that the dashboard does not have to look at multiple routes to get metrics.

File: model-mesh/overlays/odh-model-controller/rbac/role.yaml
Comment (VedantMahabaleshwarkar) [reply model-mesh/overlays/odh-model-controller/rbac/role.yaml:96]: odh-model-controller needs these because the user can choose to deploy models in any NS on the cluster (as long as modelmesh is enabled on that NS), and we need to create a CRB and a RoleBinding to make the model deployment work as intended

File: modelmesh-monitoring/base/prometheus-operator/deployment.yaml
Comment (VedantMahabaleshwarkar) [reply modelmesh-monitoring/base/prometheus-operator/deployment.yaml:1]: From a previous conversation with @anishasthana and @LaVLaS, we wanted to avoid having one operator depend on another. Hence the manual deployment

**Comment by VedantMahabaleshwarkar** [comment]
> > ```
> > * Sync model-mesh manifests from opendatahub-io/model-serving/manifests to opendatahub-io/odh-manifests/model-mesh
> > 
> > * Add model-monitoring manifests from downstream manifests to ODH manifests
> > 
> > * Delete old modelmesh basic tests that are broken
> > 
> > * Add following tests to `tests/basictests/modelmesh.sh
> > ```
> 
> Is there some circular dependency between those items ? Looks like that could be broken into separate PR for easier review.

@VannTen I wouldn't call it a circular dependency. Ideally the changes to `tests/basictests/modelmesh.sh` come after the new manifests are merged. But since we are adding new functionality, @LaVLaS wanted to include tests along with this so we can verify the PR works as intended.

File: modelmesh-monitoring/base/prometheus-operator/deployment.yaml
Comment (VannTen) [reply modelmesh-monitoring/base/prometheus-operator/deployment.yaml:1]: Several things:
1. Prometheus operator, AFAICT, is part of openshift-monitoring which is part of the openshift platform, unless I misunderstood https://docs.openshift.com/container-platform/4.10/monitoring/monitoring-overview.html So that's a dependency on the platform rather than an external operator.
2. The manifest don't have the prometheus operator  CRDs, so it won't work on cluster that indeed do not have the prometheus-operator (bare k8s for example).
3. On clusters that do have the prometheus operator, there is the potential that the two instance of the operators fight themselves trying to control CR -> chaos ensues.

IMO, using the Openshift monitoring stack (and for k8s user, declaring a dependency on prometheus operator) is better.

File: modelmesh-monitoring/base/servicemonitors/modelmesh-federated-metrics.yaml
Comment (VannTen) [reply modelmesh-monitoring/base/servicemonitors/modelmesh-federated-metrics.yaml:23]: > The plan is that the odh-dashboard will show some of these system metrics in the dashboard
Isn't the dashboard for data-scientist ? SRE and apparented are not going to look at odh-dashboard, they'll use the standard one in their cluster (grafana, most likely)

File: model-mesh/overlays/odh-model-controller/rbac/role.yaml
Comment (VannTen) [reply model-mesh/overlays/odh-model-controller/rbac/role.yaml:96]: Why do models needs access to other namespace though ? Their kinda 'user workload' right ? And you don't give full namespace access to your cluster to those.

**Comment by VannTen** [comment]
> @VannTen I wouldn't call it a circular dependency. Ideally the changes to `tests/basictests/modelmesh.sh` come after the new manifests are merged. But since we are adding new functionality, @LaVLaS wanted to include tests along with this so we can verify the PR works as intended. 

The sync part (first *) could at least be split don't you think ? 

Why not on adding the tests as part of the PR, but could you at least do
separate commits ? That's super hard to read and review as is.

File: model-mesh/overlays/odh-model-controller/rbac/role.yaml
Comment (VedantMahabaleshwarkar) [reply model-mesh/overlays/odh-model-controller/rbac/role.yaml:96]: The models don't need access to other namespaces. The ODH-model controller does. see : 
https://github.com/opendatahub-io/odh-model-controller/blob/main/controllers/inferecenceservice_serviceaccount.go#L46
https://github.com/opendatahub-io/odh-model-controller/blob/main/controllers/monitoring_controller.go#L60

File: modelmesh-monitoring/base/servicemonitors/modelmesh-federated-metrics.yaml
Comment (VedantMahabaleshwarkar) [reply modelmesh-monitoring/base/servicemonitors/modelmesh-federated-metrics.yaml:23]: Some system metrics are also planned to be available in odh-dashboard even though SRE will continue to use other dashboards. This also comes from requirements around being able to export metrics into other services (like cloudwatch for example) which requires all metrics accessible by a single prometheus instance. We want to keep manifests for upstream and downstream in sync. I think that we should keep these federated metrics so that all metrics under the model-monitoring umbrella are available via the model-monitoring prometheus

**Comment by VedantMahabaleshwarkar** [comment]
> @VannTen I wouldn't call it a circular dependency. Ideally the changes to `tests/basictests/modelmesh.sh` come after the new manifests are merged. But since we are adding new functionality, @LaVLaS wanted to include tests along with this so we can verify the PR works as intended.
> The sync part (first *) could at least be split don't you think ? Why not on adding the tests as part of the PR, but could you at least do separate commits ? That's super hard to read and review as is.

@VannTen separated out the testsuite changes in another commit

File: model-mesh/overlays/odh-model-controller/rbac/role.yaml
Comment (VannTen) [reply model-mesh/overlays/odh-model-controller/rbac/role.yaml:96]: If the model don't need access to other namespaces, odh-model-controller does not need to create clusterrolebinding. Whatever installs odh-model-controller (cluster admin, OLM) does need them, but odh-model-controller just need to **have** whatever clusterrolebinding it needs, not to **create** them by itself (which is what we're allowing here).

File: modelmesh-monitoring/base/servicemonitors/modelmesh-federated-metrics.yaml
Comment (VannTen) [reply modelmesh-monitoring/base/servicemonitors/modelmesh-federated-metrics.yaml:23]: > all metrics under the model-monitoring umbrella are available via the model-monitoring prometheus

Well that's kinda my point. Those metrics are "owned" by (-> originate from if you want) k8s components, not by model-monitoring.

File: modelmesh-monitoring/base/prometheus-operator/deployment.yaml
Comment (VedantMahabaleshwarkar) [reply modelmesh-monitoring/base/prometheus-operator/deployment.yaml:1]: 1. This part was brought up in previous discussions with @LaVLaS and @anishasthana. Eventually we want to have a unified monitoring solution for all components and then we can avoid having a prometheus operator deployed JUST for model monitoring (or any other individual component). But for the time being it was decided that it is okay to replicate what we have downstream for ODH as well. 
2. That can be added as a note in the readme, in my opinion. (that for bare k8s the operator install is needed as a prereq for model-monitoring to work)
3. This current setup has been deployed and tested downstream for at least a couple of releases, we have not observed this issue.

Comment (VannTen) [reply modelmesh-monitoring/base/prometheus-operator/deployment.yaml:1]: 3. Yeah I didn't mean it will happens, just that it's a potential source of problem.

Comment (VedantMahabaleshwarkar) [reply modelmesh-monitoring/base/prometheus-operator/deployment.yaml:1]: Explained and resolved offline with @VannTen and @anishasthana

File: modelmesh-monitoring/base/servicemonitors/modelmesh-federated-metrics.yaml
Comment (VedantMahabaleshwarkar) [reply modelmesh-monitoring/base/servicemonitors/modelmesh-federated-metrics.yaml:23]: Explained and resolved offline with @VannTen and @anishasthana

File: model-mesh/overlays/odh-model-controller/rbac/role.yaml
Comment (VedantMahabaleshwarkar) [reply model-mesh/overlays/odh-model-controller/rbac/role.yaml:96]: explained and resolved offline with @VannTen and @anishasthana

**Comment by VannTen** [comment]
/lgtm
(as far as I can tell ^)

**Review by anishasthana — APPROVED** [review]
/lgtm
/unhold

**Comment by anishasthana** [comment]
@VannTen can you dismiss your old review/ create a new review?

File: tests/setup/odh-core.yaml
Comment (LaVLaS) [inline tests/setup/odh-core.yaml:7]: Please restore the original test kfdef and insert the new modelmesh `kustomizeConfig`.  
After reviewing the CI test run failure and testing this locally, it is braking the integration tests since no other components are installed and their tests will fail.

**Review by LaVLaS — CHANGES_REQUESTED** [review]

File: model-mesh/dependencies/fvt.yaml
Comment (VedantMahabaleshwarkar) [reply model-mesh/dependencies/fvt.yaml:50]: synced the file directly from https://github.com/opendatahub-io/modelmesh-serving/blob/main/manifests/opendatahub/dependencies/fvt.yaml#L50  
But I can remove the `http://`

Comment (VedantMahabaleshwarkar) [reply model-mesh/dependencies/fvt.yaml]: removed

File: tests/setup/odh-core.yaml
Comment (VedantMahabaleshwarkar) [reply tests/setup/odh-core.yaml]: done

**Review by LaVLaS — APPROVED** [review]

**Comment by shalberd** [comment]
@LaVLaS It is great that you are adding monitorign endpoints / ports, ServiceMonitors and alerting rules. However, I think the extra monitoring stack implementation including RBAC is unnecessarily complicating things, when OKD and OCP have user namespaces monitroing built in (Monitoring for User Defined Projects). I mean, you all are at RedHat, why make things more complex than necessary? See also https://github.com/opendatahub-io/opendatahub-operator/pull/206#issuecomment-1463789365

**Comment by shalberd** [comment]
grafik

Could you keep the ServiceMonitor in Modelmesh-Monitoring base, and make prometheus and prometheus operator optional via an overlay? That way, you'd have all the metrics gathering still in there, while making it possible for users who have monitoring for user defined projects enabled to skip the prometheus part.

**Comment by LaVLaS** [comment]
Please move this discussion an open issue and not a PR that has already merged

**Comment by shalberd** [comment]
Done https://github.com/opendatahub-io/odh-manifests/issues/752

**Pull request merged**
